### DISCUSSION TITLE
Folding rows in a DataTable

### DISCUSSION THREAD

**Discussion by ajwood** [Textualize/textual#3478] [Q&A] [answered]
I'm playing around with a DataTable and wondering if there's a way I could implement collapsible rows. Specifically what I'm trying to do: when there are multiple successive rows with a repeated value in a particular column, display the first N rows in the sequence, and hide the rest; indicate somehow that there are folded rows, and provide a way to expose them.

```
folder  foo
a       1
a       2
a       3
a       4
b       5
c       6
a       7
b       8
a       9
```
would start out looking something like this (note the repeated 'a' rows are folded):
```
folder  foo
a       1
a       2
..*     ..*
b       5
c       6
a       7
b       8
a       9

```

My first thought was simply not add the collapsed rows, and only insert them when the user unfolds, however there doesn't seem to be a way to insert rows in the middle of the table, only at the bottom.

Next I tried looking for a CSS solution, add all the rows to the table upfront, but style with `display: none` when they're folded, however there doesn't seem to be a way to style specific rows like this.

Is what I'm attempting going to be possible? Or would it require deeper changes to the `DataTable` class to make it doable?

**Comment by misterek** [comment, marked as answer]
I had a similar question, I think, here: https://github.com/Textualize/textual/discussions/3067

My solution was basically:
* A hidden column for sorting.  That allowed me to sort how I wanted (In your example, it wouldn't quite work, but the hidden column would be may be "a-1", "a-2", etc..
* Manual code to fold/unfold rows.  Basically "if you click on a folded row, unfold it by adding a bunch of new rows, then re-sort based on the hidden column"

It works ok, but isn't great.  I don't bother allowing sorting by any other columns, or it would get too complicated.

**Comment by davep** [comment]
> Next I tried looking for a CSS solution, add all the rows to the table upfront, but style with display: none when they're folded, however there doesn't seem to be a way to style specific rows like this.

Rows in a `DataTable` aren't widgets so can't be styled that way.

> Is what I'm attempting going to be possible? Or would it require deeper changes to the DataTable class to make it doable?

`DataTable` doesn't support this sort of thing as it stands; yes, it would likely require deeper changes to support such a feature.
